### PR TITLE
🐛  Make all JDBC destinations (SF, RS, PG, MySQL, MSSQL, Oracle) handle wide rows by using byte-based record buffering

### PR DESCRIPTION
**Pull request by sherifnada**
## What
JDBC Destinations, when running in `INSERT` mode, typically write records as follows: 
1. Accumulate records in a buffer that can grow up to 10_000 records in size
2. Flush buffer to memory when it's full
3. Flush buffer once the sync is complete

This approach is problematic because it does not take into account how large those records are, which means wide rows can be problematic (10k 5mb rows has very different memory implications than 10k 4kb rows). This is especially exacerbated because when flushing records to the destination, the JDBC drivers needs to hold a SQL string containing all the rows in memory (i.e: `INSERT xyz INTO table` where `xyz` is the content of all the rows), so if the queue is about to write N bytes, we actually need to have a multiple of `N` bytes capacity in memory (somewhere between 2 and 3 as far as I can tell looking under the JDBC hood). 

Recently, a user was syncing very wide rows from a Postgres DB which caused this issue to appear. So this PR changes the buffering strategy to count bytes instead of number of records. 

I've confirmed this fix works for the user in question. 

## How
There are two changes being made here: 
1. keep track of the buffer size in bytes and flush when the size has reached the maximum alloted. This requires an extra serialization of the input records. This is ineffecient but didn't seem to be a huge cost when I was profiling. We definitely need more rigorous measurement of the CPU overhead of this (and we can probably even remove the need for it) but I think we should do that separately, as this fix unblocks using Snowflake destination for this category of data sources.  We can ship a more CPU efficient solution as an incremental improvement. **I've set the default buffer size to 256Mib. This way even if we hit a 3x multiple, we are using less than 1Gib of memory.** 
2. When writing records, cap all SQL `INSERT` statements to at most 10k records because some destinations like Snowflake have limitations on how many records can be written at a time (snowflake limits to 16k). 

## Recommended reading order
1. `BufferedStreamConsumer.java`
2. The rest is just adjusting everything to use the new approach

#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

### CODE REVIEW

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java [hunk 1]
```diff
@@ -141,11 +142,15 @@ protected void acceptTracked(final AirbyteMessage message) throws Exception {
         return;
       }
 
-      buffer.add(message);
-
-      if (buffer.size() == queueBatchSize) {
+      long messageSizeInBytes = Jsons.serialize(recordMessage.getData()).length() * 2; // Strings serialize to UTF-8 by default
```
Comment (tuliren) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:145]: Maybe add a TODO here about optimizing the calculation of message byte size in the future.

File: airbyte-integrations/connectors/destination-snowflake/build.gradle [hunk 2]
```diff
@@ -6,7 +6,20 @@ plugins {
...
+//            '-XX:NativeMemoryTracking=detail',
+//            "-Djava.rmi.server.hostname=localhost",
+//            '-Dcom.sun.management.jmxremote=true',
+//            '-Dcom.sun.management.jmxremote.port=6000',
+//            "-Dcom.sun.management.jmxremote.rmi.port=6000",
+//            '-Dcom.sun.management.jmxremote.local.only=false',
+//            '-Dcom.sun.management.jmxremote.authenticate=false',
+//            '-Dcom.sun.management.jmxremote.ssl=false',
+//            '-agentpath:/usr/local/YourKit-JavaProfiler-2021.3/bin/linux-x86-64/libyjpagent.so=port=10001,listen=all'
+//    ]
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-snowflake/build.gradle:21]: Nitpick. Not blocking. These commented configs about profiling are very helpful. Should move the into the `developing-locally` doc.

**Review by tuliren — APPROVED** [review]

**Comment by sherifnada** [comment]
/publish connector=connectors/destination-postgres-strict-encrypt

**Comment by sherifnada** [comment]
/publish connector=connectors/destination-mysql-strict-encrypt

**Comment by sherifnada** [comment]
/publish connector=connectors/destination-mssql-strict-encrypt

**Comment by sherifnada** [comment]
/publish connector=connectors/destination-oracle-strict-encrypt

**Comment by sherifnada** [comment]
/publish connector=connectors/destination-mssql-strict-encrypt

**Pull request merged**